### PR TITLE
drivers: dac: ad5791: Add spi mask functions, APIs

### PR DESCRIPTION
**Pull request by SaikiranGudla**
Add SPI read/write mask functions, linearity error compensation select API support.

### CODE REVIEW

**Review by amiclaus — COMMENTED** [review]
please split the patch into 2 commits, one for spi functions, one for the linearity compensation

**Review by amiclaus — COMMENTED** [review]
also check for the build issues related to the doxygen

**Comment by SaikiranGudla** [comment]
May I know if there are any inputs from the reviewers on the PR?

File: drivers/dac/ad5791/ad5791.h [hunk 1]
```diff
@@ -130,12 +131,16 @@ struct ad5791_init_param {
...
 
 /* Control Register bit Definition */
-#define AD5791_CTRL_LINCOMP(x)   (((x) & 0xF) << 6) // Linearity error compensation.
-#define AD5791_CTRL_SDODIS       (1 << 5) // SDO pin enable/disable control.
-#define AD5791_CTRL_BIN2SC       (1 << 4) // DAC register coding selection.
-#define AD5791_CTRL_DACTRI       (1 << 3) // DAC tristate control.
-#define AD5791_CTRL_OPGND        (1 << 2) // Output ground clamp control.
-#define AD5791_CTRL_RBUF         (1 << 1) // Output amplifier configuration control.
+#define AD5791_CTRL_LINCOMP_MASK	NO_OS_GENMASK(9,6)	// Linearity error compensation.
+#define AD5791_CTRL_LINCOMP(x)		(((x) & 0xF) << 6)
```
Comment (danmois) [inline drivers/dac/ad5791/ad5791.h:145]: Define values should be aligned.

File: drivers/dac/ad5791/ad5791.h [hunk 2]
```diff
@@ -130,12 +131,16 @@ struct ad5791_init_param {
...
+#define AD5791_CTRL_SDODIS_MASK     (1 << 5)			// SDO pin enable/disable control.
```
Comment (danmois) [inline drivers/dac/ad5791/ad5791.h:146]: Define values should be aligned.

Comment (danmois) [inline drivers/dac/ad5791/ad5791.h:146]: Use NO_OS_BIT for defining values where possible.

File: drivers/dac/ad5791/ad5791.c [hunk 3]
```diff
@@ -344,3 +344,73 @@ int32_t ad5791_setup(struct ad5791_dev *dev,
...
+int32_t ad5791_spi_write_mask(struct ad5791_dev *dev,
+			      uint8_t register_address,
+			      uint32_t mask,
+			      uint32_t value)
+{
+	int8_t status;
+	uint32_t reg_data;
+
+	status = ad5791_get_register_value(dev, register_address, &reg_data);
+	if (status < 0) {
```
Comment (danmois) [inline drivers/dac/ad5791/ad5791.c:372]: Checking nonzero value for status should be enough. 

No need for accolades.

Valid for other occurrences also.

File: drivers/dac/ad5791/ad5791.c [hunk 4]
```diff
@@ -344,3 +344,73 @@ int32_t ad5791_setup(struct ad5791_dev *dev,
...
+ * @param value - The register data.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+*******************************************************************************/
+int32_t ad5791_spi_write_mask(struct ad5791_dev *dev,
+			      uint8_t register_address,
+			      uint32_t mask,
+			      uint32_t value)
+{
+	int8_t status;
```
Comment (danmois) [inline drivers/dac/ad5791/ad5791.c:368]: You can specify int type for status. This is valid also for further appearances.

File: drivers/dac/ad5791/ad5791.h [hunk 5]
```diff
@@ -185,4 +190,16 @@ int32_t ad5791_soft_instruction(struct ad5791_dev *dev,
 int32_t ad5791_setup(struct ad5791_dev *dev,
 		     uint32_t setup_word);
 
+/*! SPI write to device using a mask. */
+int32_t ad5791_spi_write_mask(struct ad5791_dev *dev,
```
Comment (danmois) [inline drivers/dac/ad5791/ad5791.h:204]: Return type is int.

**Review by danmois — CHANGES_REQUESTED** [review]

File: drivers/dac/ad5791/ad5791.c (hunk 3 above)
Comment (SaikiranGudla) [reply drivers/dac/ad5791/ad5791.c:372]: The existing API functions in the drivers also use this type of return check.
Do you consider changing them as well?

Comment (danmois) [reply drivers/dac/ad5791/ad5791.c:372]: Yes.

File: drivers/dac/ad5791/ad5791.h (hunk 2 above)
Comment (SaikiranGudla) [reply drivers/dac/ad5791/ad5791.h]: Updated the code.

**Review by amiclaus — COMMENTED** [review]
your changes break the build of cn0531 which uses this driver. check the projects build log and make the adjustments the driver/project accordingly.

**Comment by SaikiranGudla** [comment]
> your changes break the build of cn0531 which uses this driver. check the projects build log and make the adjustments the driver/project accordingly.

I see that it's failing because of the macro name changed to mask subscript. Shall I add 1 more commit to rename in the project accordingly?

File: drivers/dac/ad5791/ad5791.c [hunk 6]
```diff
@@ -344,3 +344,72 @@ int32_t ad5791_setup(struct ad5791_dev *dev,
...
+	uint32_t reg_data;
+
+	status = ad5791_get_register_value(dev, register_address, &reg_data);
+	if (status)
+		return status;
+
+	reg_data &= ~mask;
+	reg_data |= value;
+
+	status = ad5791_set_register_value(dev, register_address, reg_data);
```
Comment (amiclaus) [inline drivers/dac/ad5791/ad5791.c:378]: return directly.

File: drivers/dac/ad5791/ad5791.h [hunk 7]
```diff
@@ -61,6 +61,16 @@ enum ad5791_type {
...
 };
 
+/* Possible voltage spans for linearity error compensation */
+enum ad5791_lin_comp_select {
+	AD5781_SPAN_UPTO_10V,
+	AD5791_SPAN_10V_TO_12V,
+	AD5791_SPAN_12V_TO_16V,
+	AD5791_SPAN_16V_TO_19V,
+	AD5791_SPAN_19V_TO_20V,
+	AD5781_SPAN_10V_TO_20V = 4
```
Comment (amiclaus) [inline drivers/dac/ad5791/ad5791.h:71]: if you provide value for one enum member, then all members should be set explicitly.

File: drivers/dac/ad5791/ad5791.c [hunk 8]
```diff
@@ -413,3 +413,30 @@ int ad5791_spi_read_mask(struct ad5791_dev *dev,
...
+int ad5791_set_lin_comp(struct ad5791_dev *dev,
+			enum ad5791_lin_comp_select v_span)
+{
+	int status;
+
+	if (dev->act_device == ID_AD5781 || dev->act_device == ID_AD5791) {
+		if (!v_span)
+			v_span |= NO_OS_BIT(4);
+
+		status = ad5791_spi_write_mask(dev,
```
Comment (amiclaus) [inline drivers/dac/ad5791/ad5791.c:433]: return directly.

**Comment by SaikiranGudla** [comment]
> > > your changes break the build of cn0531 which uses this driver. check the projects build log and make the adjustments the driver/project accordingly.
> > 
> > 
> > I see that it's failing because of the macro name changed to mask subscript. Shall I add 1 more commit to rename in the project accordingly?
> 
> You should add the modification in the same commit that introduces the error.

Updated the code.

File: drivers/dac/ad5791/ad5791.c [hunk 9]
```diff
@@ -344,3 +344,68 @@ int32_t ad5791_setup(struct ad5791_dev *dev,
...
+ *                          AD5791_REG_DAC          - DAC register
+ *                          AD5791_REG_CTRL         - Control register
+ *                          AD5791_REG_CLR_CODE     - Clearcode register
+ *                          AD5791_CMD_WR_SOFT_CTRL - Software control register
+ * @param mask	- The mask.
+ * @param value - Pointer to the register data.
+ *
+ * @return 0 in case of success, negative error code otherwise.
+*******************************************************************************/
+int ad5791_spi_read_mask(struct ad5791_dev *dev,
```
Comment (buha) [inline drivers/dac/ad5791/ad5791.c:396]: In my opinion, a spi_Read_mask function is useless and shouldn't be provided, as it simply does `result & mask` which is something a user could do.

This is contrary to spi_write_mask which is very useful.

Comment (SaikiranGudla) [reply drivers/dac/ad5791/ad5791.c:396]: okay, will remove the spi_read_mask function

File: drivers/dac/ad5791/ad5791.c [hunk 10]
```diff
@@ -409,3 +409,26 @@ int ad5791_spi_read_mask(struct ad5791_dev *dev,
...
+ * @brief	Set Linearity error compensation based on the reference voltage span.
+ *
+ * @param	dev		- The device structure.
+ * @param	v_span	- voltage span to set the corresponding lin_comp value.
+ * @return	0 in case of success, negative error code otherwise.
+ */
+int ad5791_set_lin_comp(struct ad5791_dev *dev,
+			enum ad5791_lin_comp_select v_span)
+{
+	if (dev->act_device == ID_AD5781 || dev->act_device == ID_AD5791) {
```
Comment (buha) [inline drivers/dac/ad5791/ad5791.c:423]: perhaps also check dev for NULL here, it could crash on the driver API user if they provide dev = NULL...

Comment (buha) [reply drivers/dac/ad5791/ad5791.c:423]: this is an issue throughout the whole file, but those instances are unrelated to your change

File: drivers/dac/ad5791/ad5791.c [hunk 11]
```diff
@@ -344,3 +344,65 @@ int32_t ad5791_setup(struct ad5791_dev *dev,
...
+ * @brief	Set Linearity error compensation based on the reference voltage span.
+ *
+ * @param	dev		- The device structure.
+ * @param	v_span	- voltage span to set the corresponding lin_comp value.
+ * @return	0 in case of success, negative error code otherwise.
+ */
+int ad5791_set_lin_comp(struct ad5791_dev *dev,
+			enum ad5791_lin_comp_select v_span)
+{
+	if(!dev)
```
Comment (amiclaus) [inline drivers/dac/ad5791/ad5791.c:394]: you can do something like:
```
{
if (!dev || (dev->act_device != ID_AD5781 && dev->act_device != ID_AD5791))
    return -EINVAL;

if (!v_span)
         v_span |= NO_OS_BIT(4);

return ad5791_spi_write_mask(dev,
				AD5791_REG_CTRL,
				AD5791_CTRL_LINCOMP_MASK,
				AD5791_CTRL_LINCOMP(v_span));
}
```
to reduce the code

**Review by amiclaus — APPROVED** [review]

**Pull request merged**